packages/core/src/extensions/Blocks/nodes/Block.ts

```diff
@@ -162,275 +109,408 @@ export const Block = Node.create<IBlock>({
-        // Try to split the current block into 2 items:
-        () => commands.splitListItem("block"),
-        // Otherwise, maybe we are in an empty list item. "Enter" should remove the list bullet
-        ({ tr, dispatch }) => {
-          const $from = tr.selection.$from;
-          if ($from.depth !== 3) {
-            // only needed at root level, at deeper levels it should be handled already by splitListItem
+        // Removes a level of nesting if the block is empty & indented, while the selection is also empty & at the start
+        // of the block.
+        () =>
+          commands.command(({ state }) => {
+            const { node, depth } = getBlockInfoFromPos(
+              state.doc,
+              state.selection.from
+            )!;
+
+            const selectionAtBlockStart =
+              state.selection.$anchor.parentOffset === 0;
+            const selectionEmpty =
+              state.selection.anchor === state.selection.head;
+            const blockEmpty = node.textContent.length === 0;
+            const blockIndented = depth > 2;
+
+            if (
+              selectionAtBlockStart &&
+              selectionEmpty &&
+              blockEmpty &&
+              blockIndented
+            ) {
+              return commands.liftListItem("block");
+            }
+
             return false;
-          }
-          const node = tr.selection.$anchor.node(-1);
-          const nodePos = tr.selection.$anchor.posAtIndex(0, -1) - 1;
-
-          if (node.type.name === "block" && node.attrs["listType"]) {
-            if (dispatch) {
-              tr.setNodeMarkup(nodePos, undefined, {
-                ...node.attrs,
-                listType: undefined,
-              });
+          }),
+        // Creates a new block and moves the selection to it if the current one is empty, while the selection is also
+        // empty & at the start of the block.
+        () =>
+          commands.command(({ state, chain }) => {
+            const { node, endPos } = getBlockInfoFromPos(
+              state.doc,
+              state.selection.from
+            )!;
+
+            const selectionAtBlockStart =
+              state.selection.$anchor.parentOffset === 0;
+            const selectionEmpty =
+              state.selection.anchor === state.selection.head;
+            const blockEmpty = node.textContent.length === 0;
+
+            if (selectionAtBlockStart && selectionEmpty && blockEmpty) {
+              const newBlockInsertionPos = endPos + 1;
+              const newBlockContentPos = newBlockInsertionPos + 2;
+
+              chain()
+                .BNCreateBlock(newBlockInsertionPos)
+                .setTextSelection(newBlockContentPos)
+                .run();
+
+              return true;
             }
-            return true;
-          }
-          return false;
-        },
-        // Otherwise, we might be on an empty line and hit "Enter" to create a new line:
-        ({ tr, dispatch }) => {
```

Comment:
Can't we use this code instead of line 414-461?

Reply:
To some extent, but we still need to make sure the selection is empty and at the start of an empty block. The enter behavior is a little different to before since before it didn't consider whether or not the selection is empty. Now, if the selection isn't empty, the content spanned by it is deleted first and then the Enter keypress is handled normally, which is what happens in most other editors. 

So for example ([] denotes selection):

12[3
qw]e

Old:
12
3
qwe

New:
12
e

This new behavior is what lines 442-460 are for.

Reply:
Basically I can use the transaction instead of the chained commands but the checks are still necessary, unless I move lines 442-460 to the top and add a selection not empty check.